File: leetcode_go/greedy_algorithms/code_360_fractional_knapsack.go
/**
 * Code360: Fractional Knapsack
 * Link: https://www.naukri.com/code360/problems/fractional-knapsack_975286
*/

package main

import (
	"fmt"
	"sort"
)

type Pair[T, U any] struct{
	Weights T
	Value U
}

func maximumValue(items []Pair[int, int], n int, w int) float64 {
	
	// original <weights, values> pairs
	fmt.Println(items)
	
	// using the sort.Slice method with a custom comparator to sort the pairs
	sort.Slice(items, func(i, j int) bool{
		return (float64(items[i].Value) / float64(items[i].Weights)) > (float64(items[j].Value) / float64(items[j].Weights))
	})

	// After sorting the items:  <weights, values> pairs
	fmt.Println(items)
	

	total_weights := 0
	total_val := 0.0

	// for index, element := range array {}
	for _, i := range items {
		if i.Weights + total_weights <= w{
			total_val += float64(i.Value)
			total_weights += i.Weights
		} else {
			total_val += (float64(i.Value) / float64(i.Weights)) * (float64(w) - float64(total_weights))
			break
		}
	}
	return total_val
}

func main(){
	items := []Pair[int, int] {{20,12}, {24,35}, {36,41}, {40,25}, {42,32}}
	fmt.Println(maximumValue(items, len(items), 100))
}

File: leetcode_problems/greedy_algorithms/code_360_fractional_knapsack.cpp
/**
 * Code360: Fractional Knapsack
 * Link: https://www.naukri.com/code360/problems/fractional-knapsack_975286
*/

#include <bits/stdc++.h>
using namespace std;

// Pair<int, int> -> Pair<weights, values>

void printVectorOfPairs(vector<pair<int,int>> &a){
    for(auto &i: a){
        cout << i.first << "," << i.second << endl;
    }
    cout << endl;
}

double maximumValue (vector<pair<int, int>>& items, int n, int w)
{
    printVectorOfPairs(items);
    std::sort(items.begin(), items.end(), [](auto &left, auto &right) {
        return ((double)left.second / (double)left.first) > ((double)right.second / (double)(right.first));
    });
    printVectorOfPairs(items);

    int total_weights = 0;

    double total_val = 0;
    
    for(auto &i: items){
        if(total_weights + i.first <= w){
            total_val += (double)i.second;
            total_weights += i.first;
            // cout << i.second << endl;
        }
        else{
            // cout << ((double)i.second / (double)i.first) * ((double)w - total_weights) << endl;
            total_val += ((double)i.second / (double)i.first) * ((double)w - total_weights);
            break;
        }
    }
    return total_val;
}

int main(){
    // vector<pair<int, int>> items = {{50,40}, {40,50}, {90,25}, {120,100}, {10,30}, {200,45}};

    vector<pair<int, int>> items = {{20,12}, {24,35}, {36,41}, {40,25}, {42,32}};

    cout << maximumValue(items, items.size(), 100) << endl;
}